Add country mapping overwrite field on geoip maxmind filter

barroca opened the pull request:
<!--
!!!ATTENTION!!!

If you are fixing *any* crash or *any* potential security issue, *do not*
open a pull request in this repo. Please report the issue via emailing
envoy-security@googlegroups.com where the issue will be triaged appropriately.
Thank you in advance for helping to keep Envoy secure.

!!!ATTENTION!!!

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)
-->

Commit Message:


[Add country map file to ovewrite country code.](https://github.com/envoyproxy/envoy/commit/e50606feee33bb3f31c2a7f2ca29252dee081dd0) 

Added :ref:`Geoip country mapping path option<envoy_v3_api_msg_extensions.geoip_providers.maxmind.v3.MaxMindConfig.country_mapping_path` to allow override of country code when needed. The country mapping file is a json containing the iso code to be mapped from:to so it is possible to override the result read from the Maxmind database. If the path is incorrect or the json is invalid, a warning will be logged, the metric country_mapping.parse_error will be incrementedand an empty map will be used.

Additional Description:
I've added a commit that changes the behaviour to throw exceptions instead of logging errors when the country mapping file is invalid. Feel free to ask for removal if you prefer to have it just sending warnings.

Risk Level:
This PR changes the maxmind geoip filter and should have low risk since it doesn't change other filter behaviours. 

Testing:
Using curl with `curl -k -H "X-Forwarded-For: 85.24.143.2, 40.0.0.1" -H "X-Real-IP: 1.1.1.10"`  might be enough to validate the configuration.

Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

phlax commented on changelogs/current.yaml:
```diff
@@ -114,6 +114,13 @@ removed_config_or_runtime:
 new_features:
+- area: geoip
+  change: |
+    Added :ref:`Geoip country mapping path option
+    <envoy_v3_api_msg_extensions.geoip_providers.maxmind.v3.MaxMindConfig.country_mapping_path` to allow override of country code when
+    needed. The country mapping file is a json containing the iso code to be mapped from:to so it is possible to override the result
+    read from the Maxmind database. If the path is incorrect or the json is invalid, a warning will be logged, the metric
+    country_mapping.parse_error will be incremented and an empty map will be used.
```
```suggestion
    Added :ref:`Geoip country mapping path option
    <envoy_v3_api_msg_extensions.geoip_providers.maxmind.v3.MaxMindConfig.country_mapping_path` to allow override of country code when
    needed. The country mapping file is JSON containing the ISO code to be mapped from:to so it is possible to override the result
    read from the Maxmind database. If the path is incorrect or the JSON is invalid, a warning will be logged, the metric
    ``country_mapping.parse_error`` will be incremented and an empty map will be used.
```

phlax reviewed (commented):
thanks for working on this @barroca - minor changelog nits ...

phlax commented on changelogs/current.yaml:
```diff
@@ -114,6 +114,13 @@ removed_config_or_runtime:
+- area: geoip
+  change: |
+    Added :ref:`Geoip country mapping path option
+    <envoy_v3_api_msg_extensions.geoip_providers.maxmind.v3.MaxMindConfig.country_mapping_path` to allow override of country code when
```
```suggestion
    <envoy_v3_api_field_extensions.geoip_providers.maxmind.v3.MaxMindConfig.country_mapping_path` to allow override of country code when
```

phlax commented on api/envoy/extensions/geoip_providers/maxmind/v3/maxmind.proto:
```diff
@@ -39,4 +40,11 @@ message MaxMindConfig {
       [(validate.rules).message = {required: true}];
+
+  // Optional full file path to the config country mappings overrides, e.g. /etc/country-mapping.conf.
+  // File expected to have .conf extension. File should be a json with the following format:
+  // {
+  //   "country_code": "country_code" # comment
+  // }
```
```suggestion
  // The file is expected to have a ``.conf`` extension, and should be JSON with the following format:
  //
  // .. code-block::json
  //   {
  //     "country_code": "country_code" # comment
  //   }
  //
```

not 100% if this will resolve the current docs build issues but i think we want to do this anyway

phlax commented on api/envoy/extensions/geoip_providers/maxmind/v3/maxmind.proto:
```diff
@@ -39,4 +40,11 @@ message MaxMindConfig {
   common.v3.CommonGeoipProviderConfig common_provider_config = 4
       [(validate.rules).message = {required: true}];
+
+  // Optional full file path to the config country mappings overrides, e.g. /etc/country-mapping.conf.
```
```suggestion
  // Optional full file path to the config country mappings overrides, e.g. ``/etc/country-mapping.conf``.
```

phlax reviewed (commented):
some fixes/suggestions for docs

phlax commented:
/docs

barroca commented:
/docs

phlax commented on api/envoy/extensions/geoip_providers/maxmind/v3/maxmind.proto:
```diff
@@ -39,4 +40,15 @@ message MaxMindConfig {
+  // .. code-block::json
+  //   {
+  //     "country_code": "country_code" # comment
+  //   }
```
```suggestion
  //   }
  //
```

phlax commented:
@barroca the docs will render once they are able to - theres still an issue with the changelog link

phlax commented on changelogs/current.yaml:
```diff
@@ -114,6 +114,13 @@ removed_config_or_runtime:
+- area: geoip
+  change: |
+    Added :ref:`Geoip country mapping path option
+    <envoy_v3_api_field_extensions.geoip_providers.maxmind.v3.MaxMindConfig.country_mapping_path` to allow override of country code when
```
```suggestion
    <envoy_v3_api_field_extensions.geoip_providers.maxmind.v3.MaxMindConfig.country_mapping_path>` to allow override of country code when
```

ravenblackx commented on api/envoy/extensions/geoip_providers/maxmind/v3/maxmind.proto:
```diff
@@ -39,4 +40,16 @@ message MaxMindConfig {
+  //
+  // .. code-block::json
+  //   {
+  //     "country_code": "country_code" # comment
```
Probably worth clarifying that it's not actually json because json doesn't support comments.

I might also suggest using `//` comments rather than `#` comments, adding an optional ignore_comments argument in the json loader (to be used [here](https://github.com/envoyproxy/envoy/blob/main/source/common/json/json_internal.cc#L728) as the argument to [sax_parse](https://json.nlohmann.me/api/basic_json/sax_parse/)) and supporting comments that way ([nlohmann's blurb about comments](https://json.nlohmann.me/features/comments/))

That would be better tested than the regex which would, for example, have a surprising outcome if someone were to try to use some sort of meta country codes like
```
{
  "#X": "boop"
}
```
(The hacky regex method as currently coded will transform this into invalid json.)

phlax replied:
seems the json is not rendering - im guessing the comment is why (https://storage.googleapis.com/envoy-pr/bdfa297/docs/api-v3/extensions/geoip_providers/maxmind/v3/maxmind.proto.html)

if the comment is necessary you can add `:force:` to the json code-block, otherwise maybe just remove the comment

also wondering if the json code-block needs a space - ie `.. code-block:: json`

barroca replied:
made some changes on the parser

phlax replied:
JSON doesnt support any comments - so this is still failing

as said above - if comments really _should_ be included we can force it - but its indicating that (most) json parsers are not going to work with comments included

ravenblackx commented on source/extensions/geoip_providers/maxmind/geoip_provider.cc:
```diff
@@ -111,6 +126,62 @@ MaxmindDbPtr GeoipProvider::initMaxMindDb(const absl::optional<std::string>& db_
+    });
+  }
+  END_TRY
+  catch (Json::Exception&) {
```
I think `END_TRY CATCH` might be the expected thing here, in support of a build with disabled exceptions. ([example](https://sourcegraph.com/github.com/envoyproxy/envoy@4994e59af25898d56f496fece3ed24dacbdfa87b/-/blob/source/common/router/header_parser_utils.cc?L54))

Also, including the message from the exception in the rethrown exception would probably be helpful to the user, rather than just telling them the entire file is wrong.

ravenblackx commented on test/extensions/geoip_providers/maxmind/test_data/country_mapping.conf:
```diff
@@ -0,0 +1,6 @@
+{
+  "PF": "AU", // French Polynesia
+  "PT-30": "MA", // Madeira
+  "SE": "XX" // Ovewrite for testing country mapping
```
Overwrite

ravenblackx reviewed (dismissed):
This looks okay to me except the one super-nit comment typo. (And some bad test style things in the json test but they're consistent with what's already there so I'm not gonna ask you to fix them.)

nezdolik commented on source/extensions/geoip_providers/maxmind/geoip_provider.cc:
```diff
@@ -27,6 +37,10 @@ GeoipProviderConfig::GeoipProviderConfig(
                                                  : absl::nullopt),
       anon_db_path_(!config.anon_db_path().empty() ? absl::make_optional(config.anon_db_path())
                                                    : absl::nullopt),
+      // Path is hardcoded since we don't have the field on the OSS version of the proto.
```
nit: remove non relevant comment

nezdolik commented on source/extensions/geoip_providers/maxmind/geoip_provider.cc:
```diff
@@ -111,6 +126,55 @@ MaxmindDbPtr GeoipProvider::initMaxMindDb(const absl::optional<std::string>& db_
+      }
+      parseCountryMapping(file_or_error.value(), country_mapping_path.value(), country_mapping);
+    } else {
+      config_->incCountryMappingFileParseError();
```
practically, this is not file parse error, but rather file does not exist/could not be resolved by filesystem object.

nezdolik commented on source/extensions/geoip_providers/maxmind/geoip_provider.cc:
```diff
@@ -150,6 +214,12 @@ void GeoipProvider::lookupInCityDb(
+          if (it != lookup_result.end()) {
+            country = it->second;
+          }
+          lookup_result[config_->countryHeader().value()] = getCountryMapping(country);
```
highly likely this assignment operator will copy the string value of `getCountryMapping(country)` call, you could use move semantics (capture getCountryMapping(country) by rvalue) instead to avoid extra copy: https://cplusplus.com/reference/string/string/operator=/

nezdolik commented on test/extensions/geoip_providers/maxmind/config_test.cc:
```diff
@@ -353,6 +402,34 @@ TEST(MaxmindProviderConfigTest, DifferentProviderInstancesForDifferentProtoConfi
+      geo_headers_to_add:
+        city: "x-geo-city"
+    city_db_path: %s
+    country_mapping_path: "/usr/share/geoip2Gcs/wrong_path.conf"
```
since this is optional config param, we should have a test where this field is not configured.

nezdolik commented on test/extensions/geoip_providers/maxmind/config_test.cc:
```diff
@@ -331,19 +370,29 @@ TEST(MaxmindProviderConfigTest, DifferentProviderInstancesForDifferentProtoConfi
+
+  auto api = Api::createApiForTest();
+
+  Stats::IsolatedStoreImpl stats_store_;
+  Stats::ScopeSharedPtr scope_{stats_store_.createScope("")};
```
these variables are local to a single test code, the names should not have underscores in the end.

nezdolik commented on test/extensions/geoip_providers/maxmind/geoip_provider_test.cc:
```diff
@@ -98,6 +108,34 @@ TEST_F(GeoipProviderTest, ValidConfigCityAndIspDbsSuccessfulLookup) {
   expectStats("isp_db");
 }
 
+TEST_F(GeoipProviderTest, ValidConfigCityAndIspDbsSuccessfulLookupWithoutResultDontCauseCrash) {
```
could you elaborate more/add explanatory comment how this test relates to newly added country mapping feature? E.g. why could there be a crash

nezdolik commented on test/extensions/geoip_providers/maxmind/geoip_provider_test.cc:
```diff
@@ -265,6 +304,74 @@ TEST_F(GeoipProviderTest, ValidConfigCityMultipleLookups) {
   expectStats("city_db", 2, 2);
 }
 
+TEST_F(GeoipProviderTest, ValidCountryMapping) {
```
nit: rename test to smth like `ValidCountryMappingCountryValueOverriden` etc

nezdolik replied:
also consider splitting 2 cases in this test into 2 tests

nezdolik commented on test/extensions/geoip_providers/maxmind/geoip_provider_test.cc:
```diff
@@ -287,6 +394,40 @@ TEST_F(GeoipProviderDeathTest, GeoDbNotSetForConfiguredHeader) {
                "performing lookups");
 }
 
+TEST_F(GeoipProviderDeathTest, CountryMappingFileStatusNotOk) {
```
i may have missed it, but is there is a test when file path does not exist?

nezdolik reviewed (commented):
Thank you for the contribution! Please update integration test as well

ravenblackx commented:
Now that nezdolik has the comments I'm going to assign it over. :)

nezdolik commented:
Apologies for delayed review, i have been swamped past weeks.

barroca commented:
@nezdolik I addressed your comments, but right now I'm having trouble identifying the leak that I'm getting from the presubmit  job https://dev.azure.com/cncf/envoy/_build/results?buildId=163920&view=logs&j=1439b9f7-a348-5b50-b5fe-ea612ea91241&t=1002ac43-da84-5fae-70b2-98833b702d09

I'm still investigating, but if you have any clue let me know.

nezdolik commented on source/extensions/geoip_providers/maxmind/geoip_provider.cc:
```diff
@@ -88,29 +103,88 @@ GeoipProvider::~GeoipProvider() {
   ENVOY_LOG(debug, "Shutting down Maxmind geolocation provider");
   if (city_db_) {
     MMDB_close(city_db_.get());
+    city_db_ = nullptr;
```
why is this change needed? Database pointer is a unique smart pointer owned by provider instance, memory will be released by smart pointer on geoip provider destruction.

nezdolik commented on source/extensions/geoip_providers/maxmind/geoip_provider.cc:
```diff
@@ -88,29 +103,88 @@ GeoipProvider::~GeoipProvider() {
     MMDB_close(anon_db_.get());
+    anon_db_ = nullptr;
+  }
+  if (country_mapping_) {
```
why this is `if` block is needed?

barroca replied:
country_mapping_ is a pointer as anon_db_ and others. keeping the same standard

nezdolik replied:
country_mapping_ is smart pointer to C++ `absl::map`, there is no need to clean it within the destructor of the owning object (provider). While all db types are smart pointers to C struct types, in C there are no destructors and maxmind library recommends to call `MMDB_close` to release dynamically allocated memory.

nezdolik commented on source/extensions/geoip_providers/maxmind/geoip_provider.cc:
```diff
@@ -88,29 +103,88 @@ GeoipProvider::~GeoipProvider() {
   if (db_path) {
-    MMDB_s maxmind_db;
-    int result_code = MMDB_open(db_path.value().c_str(), MMDB_MODE_MMAP, &maxmind_db);
+    MMDB_s* maxmind_db = new MMDB_s;
```
please revert to previous version, Envoy uses smart pointers for memory management, this is the reason you got a memory leak.
